src/Microsoft.TestPlatform.Common/ExtensionDecorators/SerializeTestRunDecorator.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Xml.Linq;
+
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
+namespace Microsoft.VisualStudio.TestPlatform.Common.ExtensionDecorators;
+
+internal class SerializeTestRunDecorator : ITestExecutor, ITestExecutor2, IDisposable
+{
+    private readonly SemaphoreSlim _runSequentialEvent = new(1);
+
+    public ITestExecutor OriginalTestExecutor { get; private set; }
+
+    public SerializeTestRunDecorator(ITestExecutor originalTestExecutor)
+    {
+        OriginalTestExecutor = originalTestExecutor;
+    }
+
+    public void RunTests(IEnumerable<TestCase>? tests, IRunContext? runContext, IFrameworkHandle? frameworkHandle)
+    {
+        if (IsSerializeTestRunEnabled(runContext))
+        {
+            EqtTrace.Info($"SerializeTestRunDecorator: Test cases will run sequentially");
+            if (tests is null)
+            {
+                return;
+            }
+
+            foreach (TestCase testToRun in tests)
+            {
+                _runSequentialEvent.Wait();
+                OriginalTestExecutor.RunTests(new List<TestCase> { testToRun }, runContext, new SerializeTestRunDecoratorFrameworkHandle(frameworkHandle!, _runSequentialEvent));
```

Comment:
Is it ok to force passing `null` for the `frameworkHandle`?

Reply:
Actually "cannot be null", the decorator is not interfering with the current nullability rule.